pkg/cover/backend/elf.go

```diff
@@ -40,85 +253,130 @@ func elfReadSymbols(module *Module, info *symbolInfo) ([]*Symbol, error) {
-			log.Logf(0, "ignoring module %v without DEBUG_INFO", module.Name)
-			return nil, nil, nil
-		}
-		return nil, nil, fmt.Errorf("failed to parse DWARF: %v (set CONFIG_DEBUG_INFO=y on linux)", err)
-	}
-
-	var pcFix pcFixFn
-	if kaslr {
-		pcFix = func(r [2]uint64) ([2]uint64, bool) {
-			if r[0] >= r[1] || r[0] < text.Addr || r[1] > text.Addr+text.Size {
-				// Linux kernel binaries with CONFIG_RANDOMIZE_BASE=y are strange.
-				// .text starts at 0xffffffff81000000 and symbols point there as
-				// well, but PC ranges point to addresses around 0.
-				// So try to add text offset and retry the check.
-				// It's unclear if we also need some offset on top of text.Addr,
-				// it gives approximately correct addresses, but not necessary
-				// precisely correct addresses.
-				r[0] += text.Addr
-				r[1] += text.Addr
-				if r[0] >= r[1] || r[0] < text.Addr || r[1] > text.Addr+text.Size {
-					return r, true
+		return nil, err
+	}
+	var symbols []*Symbol
+	for _, ds := range allDwarfFunctions {
+		if len(ds.Ranges) == 0 {
+			continue
+		}
+		start := ds.Ranges[0][0]
+		end := ds.Ranges[len(ds.Ranges)-1][1]
+		var ranges [][2]uint64
+		for _, r := range ds.Ranges {
+			ranges = append(ranges, [2]uint64{
+				r[0] + module.Addr,
+				r[1] + module.Addr,
+			})
+		}
+		flag := start >= text.Addr && end <= text.Addr+text.Size
+		if flag {
+			start1 := start + module.Addr
+			end1 := end + module.Addr
+			symbols = append(symbols, &Symbol{
+				Module: module,
+				ObjectUnit: ObjectUnit{
+					Name: ds.Name,
+				},
+				Start:  start1,
+				End:    end1,
+				Ranges: ranges,
+				Inline: ds.Inline,
+				Unit: &CompileUnit{
+					Module: module,
+					ObjectUnit: ObjectUnit{
+						Name: ds.DeclFile,
+					},
+				},
+			})
+		}
+	}
+	return symbols, nil
+}
+
+func elfReadSymbolsFromDwarf(debugInfo *dwarf.Data, compileUnits []*DwarfCompileUnit) ([]*DwarfFunction, error) {
+	var allSubprograms []*DwarfFunction
+	var allInlineSubroutines []*DwarfFunction
+	for _, cu := range compileUnits {
+		errc := make(chan error, 1)
+		go func(cu *DwarfCompileUnit) {
+			subprograms, err := readAllSubprograms(debugInfo, cu)
+			if err != nil {
+				errc <- err
+				return
+			}
+			allSubprograms = append(allSubprograms, subprograms...)
+			inlineSubprograms := make(map[dwarf.Offset]*DwarfFunction)
+			for _, sp := range subprograms {
+				if sp.Inline {
+					inlineSubprograms[sp.Offset] = sp
+				}
+			}
+			for _, sp := range subprograms {
+				if sp.Inline {
+					continue
+				}
+				erric := make(chan error, 1)
+				go func(sp *DwarfFunction) {
+					inlineSubroutines, err := readAllInlinedSubroutines(debugInfo, inlineSubprograms, sp)
+					if err != nil {
+						erric <- err
+						return
+					}
+					allInlineSubroutines = append(allInlineSubroutines, inlineSubroutines...)
+					erric <- nil
+				}(sp)
+				if err := <-erric; err != nil {
+					errc <- err
 				}
 			}
-			return r, false
+			errc <- nil
+		}(cu)
+		if err := <-errc; err != nil {
```

Comment:
Actually, I don't see the value of using goroutines here at all - you anyway hang here on the error channel and will only continue to the next goroutine once the previous one has finished execution.

Did I miss something?

Reply:
@dvyukov I see the same is done in the existing code in `makeDWARFUnsafe`. Was it done to keep possible segfault crashes off the main thread?

Reply:
Better let Dmitry to comment why (it puzzles me too), if I put
		if err := <-errc; err != nil {
			return nil, err
		}
after the outer for loop, then I get wrong functions.

And if without the outer and inner go routine, performance drops ~ 20x.